### ISSUE TITLE
server: Incorrect logit process for mixed CPU/GPU collection in StaticWarper - Logit process

### ISSUE THREAD

**Issue by michaelfeil** [huggingface/text-generation-inference#680]
### System Info

This is not exactly a bug, that should occur in production with the current code as it is.

### Information

- [ ] Docker
- [X] The CLI directly

### Tasks

- [ ] An officially supported command
- [X] My own modifications

## Reproduce:
Modify causal_lm.py:
- disable in __init__.py that model explicitly moved to CPU -> device cpu https://github.com/huggingface/text-generation-inference/blob/1da642bd0e6de28ef499f17cd226264f3ccdc824/server/text_generation_server/models/causal_lm.py#L477-L488
- run the santacoder_model unit tests, with a modified temperature or top_k to make sure that non-greedy sampling is selected.

## Observation
- Here the returned logits will be torch.Tensor(cpu, float32) instead of torch.Tensor(cuda, float16)  https://github.com/huggingface/text-generation-inference/blob/1da642bd0e6de28ef499f17cd226264f3ccdc824/server/text_generation_server/models/causal_lm.py#L532

- This causes a silent error here in next token selection: https://github.com/huggingface/text-generation-inference/blob/1da642bd0e6de28ef499f17cd226264f3ccdc824/server/text_generation_server/models/causal_lm.py#L576
The first generated token is repeated. I modified the logits to be esentially a "one-hot-tensor", setting the argmax() token to +10 and the others to -10, still getting a different token selected by the next_token_selection function.

I suspect the bug here.
https://github.com/huggingface/text-generation-inference/blob/1da642bd0e6de28ef499f17cd226264f3ccdc824/server/text_generation_server/utils/logits_process.py#L45

## Proposed changes
I would modify https://github.com/huggingface/text-generation-inference/blob/1da642bd0e6de28ef499f17cd226264f3ccdc824/server/text_generation_server/utils/logits_process.py#L45 to check if the scores tensor is not on cpu, instead of relying that the tensor is on gpu if gpu is available.

```python
# something like:
if torch.cuda.is_available() and scores.device != "cpu"
```

## Effect:
Because the code is relying on the fact, that all models are moved to GPU if cuda is available, the code works as is in production. Also the code works if no temperature sampling is used by greedy sampling instead.
However it would save significant debugging time (as for me 2-3h) 

### Expected behavior

The next_token selection should have no side-effects, and operate only on its configuration and the logits tensor.

**Comment by njhill** [comment]
I think it would be best to just change that line to
```python
if scores.device.type == "cuda":
```

@michaelfeil maybe you could open a PR for consideration?

**Comment by michaelfeil** [comment]
@njhill Added this actually to the other PR - this seems to fix it.
